<!DOCTYPE html>

<html>

<head>

<meta charset="utf-8" />
<meta name="generator" content="pandoc" />
<meta http-equiv="X-UA-Compatible" content="IE=EDGE" />

<meta name="viewport" content="width=device-width, initial-scale=1" />

<meta name="author" content="Corey Broeckling" />

<meta name="date" content="2025-08-22" />

<title>pubchem.bio</title>

<script>// Pandoc 2.9 adds attributes on both header and div. We remove the former (to
// be compatible with the behavior of Pandoc < 2.8).
document.addEventListener('DOMContentLoaded', function(e) {
  var hs = document.querySelectorAll("div.section[class*='level'] > :first-child");
  var i, h, a;
  for (i = 0; i < hs.length; i++) {
    h = hs[i];
    if (!/^h[1-6]$/i.test(h.tagName)) continue;  // it should be a header h1-h6
    a = h.attributes;
    while (a.length > 0) h.removeAttribute(a[0].name);
  }
});
</script>

<style type="text/css">
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
span.underline{text-decoration: underline;}
div.column{display: inline-block; vertical-align: top; width: 50%;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
</style>



<style type="text/css">
code {
white-space: pre;
}
.sourceCode {
overflow: visible;
}
</style>
<style type="text/css" data-origin="pandoc">
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { display: inline-block; text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
{ counter-reset: source-line 0; }
pre.numberSource code > span
{ position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
{ content: counter(source-line);
position: relative; left: -1em; text-align: right; vertical-align: baseline;
border: none; display: inline-block;
-webkit-touch-callout: none; -webkit-user-select: none;
-khtml-user-select: none; -moz-user-select: none;
-ms-user-select: none; user-select: none;
padding: 0 4px; width: 4em;
color: #aaaaaa;
}
pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa; padding-left: 4px; }
div.sourceCode
{ }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
code span.al { color: #ff0000; font-weight: bold; } 
code span.an { color: #60a0b0; font-weight: bold; font-style: italic; } 
code span.at { color: #7d9029; } 
code span.bn { color: #40a070; } 
code span.bu { color: #008000; } 
code span.cf { color: #007020; font-weight: bold; } 
code span.ch { color: #4070a0; } 
code span.cn { color: #880000; } 
code span.co { color: #60a0b0; font-style: italic; } 
code span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } 
code span.do { color: #ba2121; font-style: italic; } 
code span.dt { color: #902000; } 
code span.dv { color: #40a070; } 
code span.er { color: #ff0000; font-weight: bold; } 
code span.ex { } 
code span.fl { color: #40a070; } 
code span.fu { color: #06287e; } 
code span.im { color: #008000; font-weight: bold; } 
code span.in { color: #60a0b0; font-weight: bold; font-style: italic; } 
code span.kw { color: #007020; font-weight: bold; } 
code span.op { color: #666666; } 
code span.ot { color: #007020; } 
code span.pp { color: #bc7a00; } 
code span.sc { color: #4070a0; } 
code span.ss { color: #bb6688; } 
code span.st { color: #4070a0; } 
code span.va { color: #19177c; } 
code span.vs { color: #4070a0; } 
code span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } 
</style>
<script>
// apply pandoc div.sourceCode style to pre.sourceCode instead
(function() {
  var sheets = document.styleSheets;
  for (var i = 0; i < sheets.length; i++) {
    if (sheets[i].ownerNode.dataset["origin"] !== "pandoc") continue;
    try { var rules = sheets[i].cssRules; } catch (e) { continue; }
    var j = 0;
    while (j < rules.length) {
      var rule = rules[j];
      // check if there is a div.sourceCode rule
      if (rule.type !== rule.STYLE_RULE || rule.selectorText !== "div.sourceCode") {
        j++;
        continue;
      }
      var style = rule.style.cssText;
      // check if color or background-color is set
      if (rule.style.color === '' && rule.style.backgroundColor === '') {
        j++;
        continue;
      }
      // replace div.sourceCode by a pre.sourceCode rule
      sheets[i].deleteRule(j);
      sheets[i].insertRule('pre.sourceCode{' + style + '}', j);
    }
  }
})();
</script>




<style type="text/css">body {
background-color: #fff;
margin: 1em auto;
max-width: 700px;
overflow: visible;
padding-left: 2em;
padding-right: 2em;
font-family: "Open Sans", "Helvetica Neue", Helvetica, Arial, sans-serif;
font-size: 14px;
line-height: 1.35;
}
#TOC {
clear: both;
margin: 0 0 10px 10px;
padding: 4px;
width: 400px;
border: 1px solid #CCCCCC;
border-radius: 5px;
background-color: #f6f6f6;
font-size: 13px;
line-height: 1.3;
}
#TOC .toctitle {
font-weight: bold;
font-size: 15px;
margin-left: 5px;
}
#TOC ul {
padding-left: 40px;
margin-left: -1.5em;
margin-top: 5px;
margin-bottom: 5px;
}
#TOC ul ul {
margin-left: -2em;
}
#TOC li {
line-height: 16px;
}
table {
margin: 1em auto;
border-width: 1px;
border-color: #DDDDDD;
border-style: outset;
border-collapse: collapse;
}
table th {
border-width: 2px;
padding: 5px;
border-style: inset;
}
table td {
border-width: 1px;
border-style: inset;
line-height: 18px;
padding: 5px 5px;
}
table, table th, table td {
border-left-style: none;
border-right-style: none;
}
table thead, table tr.even {
background-color: #f7f7f7;
}
p {
margin: 0.5em 0;
}
blockquote {
background-color: #f6f6f6;
padding: 0.25em 0.75em;
}
hr {
border-style: solid;
border: none;
border-top: 1px solid #777;
margin: 28px 0;
}
dl {
margin-left: 0;
}
dl dd {
margin-bottom: 13px;
margin-left: 13px;
}
dl dt {
font-weight: bold;
}
ul {
margin-top: 0;
}
ul li {
list-style: circle outside;
}
ul ul {
margin-bottom: 0;
}
pre, code {
background-color: #f7f7f7;
border-radius: 3px;
color: #333;
white-space: pre-wrap; 
}
pre {
border-radius: 3px;
margin: 5px 0px 10px 0px;
padding: 10px;
}
pre:not([class]) {
background-color: #f7f7f7;
}
code {
font-family: Consolas, Monaco, 'Courier New', monospace;
font-size: 85%;
}
p > code, li > code {
padding: 2px 0px;
}
div.figure {
text-align: center;
}
img {
background-color: #FFFFFF;
padding: 2px;
border: 1px solid #DDDDDD;
border-radius: 3px;
border: 1px solid #CCCCCC;
margin: 0 5px;
}
h1 {
margin-top: 0;
font-size: 35px;
line-height: 40px;
}
h2 {
border-bottom: 4px solid #f7f7f7;
padding-top: 10px;
padding-bottom: 2px;
font-size: 145%;
}
h3 {
border-bottom: 2px solid #f7f7f7;
padding-top: 10px;
font-size: 120%;
}
h4 {
border-bottom: 1px solid #f7f7f7;
margin-left: 8px;
font-size: 105%;
}
h5, h6 {
border-bottom: 1px solid #ccc;
font-size: 105%;
}
a {
color: #0033dd;
text-decoration: none;
}
a:hover {
color: #6666ff; }
a:visited {
color: #800080; }
a:visited:hover {
color: #BB00BB; }
a[href^="http:"] {
text-decoration: underline; }
a[href^="https:"] {
text-decoration: underline; }

code > span.kw { color: #555; font-weight: bold; } 
code > span.dt { color: #902000; } 
code > span.dv { color: #40a070; } 
code > span.bn { color: #d14; } 
code > span.fl { color: #d14; } 
code > span.ch { color: #d14; } 
code > span.st { color: #d14; } 
code > span.co { color: #888888; font-style: italic; } 
code > span.ot { color: #007020; } 
code > span.al { color: #ff0000; font-weight: bold; } 
code > span.fu { color: #900; font-weight: bold; } 
code > span.er { color: #a61717; background-color: #e3d2d2; } 
</style>




</head>

<body>




<h1 class="title toc-ignore">pubchem.bio</h1>
<h4 class="author">Corey Broeckling</h4>
<h4 class="date">2025-08-22</h4>



<div id="background" class="section level2">
<h2>Background</h2>
<p>Metabolomics approaches utilize analytical instrumentation to detect
and quantify small molecules. The chemical space covered by metabolomics
is pontentially massive, with more than 100 million structures in
Pubchem. Efforts to curate custom metabolome databases, such as the <a href="https://www.hmdb.ca/">Human Metabolome Database</a> are incredibly
valuable, enabling annotation efforts to focus on only those chemicals
most likely to be found in the samples being analyzed. Metabolomics,
however, is well suited the analysis of metabolomes from <em>any</em>
species, and there are very few taxonomically informed databases
available.</p>
<p>Pubchem, which is a massive repository of chemical structures, is
also extremely rich in chemical metadata. Several biologicallly-focused
metabolomic databases have deposited their structures data into PubChem,
thousands of metabolic pathways are present, and there are rich
taxonomic-structure pairs through taxonomy-informed natural products
database efforts.</p>
<p>The pubchem.bio package enables researchers to</p>
<ul>
<li>Download metabolomics-centric subset of PubChem onto their local
computer</li>
<li>Build a metabolomic structured database (data.frame) of all
biological compounds in PubChem</li>
<li>Define a core biological metabolome, comprising metabolites
plausibly found in any species</li>
<li>Develop custom metabolomic structure databases using selected or all
available taxonomic data in PubChem.</li>
</ul>
<p>As such, this package will help to streamline metabolomic database
creation, increase the accuracy of annotation efforts by</p>
</div>
<div id="installation" class="section level2">
<h2>Installation</h2>
<div class="sourceCode" id="cb1"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb1-1"><a href="#cb1-1" tabindex="-1"></a><span class="fu">install.packages</span>(pubchem.bio, <span class="at">dependencies =</span> <span class="cn">TRUE</span>)</span>
<span id="cb1-2"><a href="#cb1-2" tabindex="-1"></a><span class="fu">library</span>(pubchem.bio)</span></code></pre></div>
</div>
<div id="system-requirements" class="section level2">
<h2>System requirements</h2>
<p>PubChem is a very large database. The pubchem.bio package does not
download <em>ALL</em> of PubChem. It does however download every
structure, inchikey, smiles, and several other basic descriptors. As
such, while the package is designed to minimize computational overhead,
you will still want to have sufficient disk space and RAM to ensure that
processing doesn’t bog down during parsing and organizing files.</p>
<p>rcdk is an R package which enables calculations of chemical
properties. it requires 64 bit Java to be installed on your computer. If
you do not have 64 bit Java installed, you will receive an error message
upon trying to load pubchem.bio or rcdk, something to the effect that
‘Java not found’. Most computers do not have 64 bit Java installed by
default, so you will likely need to install this manually using the
manual Java instructions - just search ‘install 64 bit java’ in your
favorite internet browser for the latest hyperlink. <em>Then</em> open
Rstudio and install pubchem.bio.</p>
<p>It is recommended that you perform these functions on a computer with
at least <strong>200 GB local SSD</strong> disk space, since a great
deal of reading and writing of data will be occurring. Further, during
processing, there will be large files that are being handled in memory,
so it is recommended that you utilize a computer with at least
<strong>64 GB of RAM</strong>. Parallel processing is enabled through
the doParallel and foreach packages, which will speed up processing
time, but the will still take several hours. Do note that parallel
processing results in increased RAM consumption. If your run out of RAM,
the process will become extremely slow and could crash R/Rstudio. If you
are RAM limited, you are probably going to have more success using fewer
threads. The download process performance is also going to be dependent
on network stability and speed.</p>
<p>Running the program on a windows 11 PC, with 64 GB RAM, a local SSD
hard drive, and setting threads = 8 in all functions, tyou can expect
download, organization, and metabolome generation to take about 4 hours
of computer time on a modern laptop/desktop. <em>Once you have
everything set up, each new metabolome generated can be created in
minutes.</em></p>
</div>
<div id="running-the-code" class="section level2">
<h2>Running the code</h2>
<p>This isn’t a complex package, there are relatively few functions. All
functions start with ‘get’, ‘build’, or ‘export’. ‘get’ functions
download data to your local drive, ‘build’ functions parse and organize
the downloaded data to turn it into something more useful for
metabolomic scientists, and ‘export’ functions write these metabolome
datasets to file for use in other programs.</p>
<div id="downloading-all-the-pubchem-we-need" class="section level3">
<h3>Downloading all the PubChem we need</h3>
<p>First thing we need to do is get PubChem data local to your
computer.</p>
<div id="note-that-running-this-line-of-code-may-occupy-your-r-session-for-2-3-few-hours" class="section level6">
<h6>note that running this line of code may occupy your R session for
2-3 few hours</h6>
<div class="sourceCode" id="cb2"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb2-1"><a href="#cb2-1" tabindex="-1"></a>pubchem.bio<span class="sc">::</span><span class="fu">get.pubchem.ftp</span>(<span class="at">pc.directory =</span><span class="st">&quot;C:/Temp/20250703&quot;</span>)</span></code></pre></div>
<p>What just happened? We have downloaded a bunch of data, mostly from
the PubChem FTP site. These downloaded compressed files were
decompressed, and we extracted only the most relevant bits of data from
them, saving those relevant bits as data.tables in .Rdata format. The
temporary files were removed after processing is done (assuming
rm.tmp.files = TRUE). We did not build a database, we simply moved all
the data we want from remote to local, and discarded data we have no
need for to keep from cluttering up our hard drive and RAM.</p>
</div>
</div>
<div id="building-metabolite-cid---lowest-common-ancestor-lca-relationships" class="section level3">
<h3>Building metabolite (CID) - lowest common ancestor (LCA)
relationships</h3>
<p>Our next step needs to be further organization of taxonomy data. This
function is separated out as it is relatively computationally intensive,
taking a good deal of time. This will be done using the ‘build.cid.lca’
function. The downloaded data is currently in a format containing three
columns, a taxonomy ID, a PubChem CID, and the source of the
relationship information. This is a nice, clean format. However,
taxonomy is hierarchical, and the pubchem.bio package aims to use that
hierarchy. The ‘build.taxid.lca function organizes the chemical
relationships into the taxonomic hierarchy, so that we can utilize this
hierarchy in building taxon-specific metabolomics databases. The
principle on which the process is based is the ’lowest common ancestor’
approach. As an example, the lowest common ancestor, from a taxonomy
perspective, of the two species <a href="https://www.ncbi.nlm.nih.gov/Taxonomy/Browser/wwwtax.cgi?lvl=0&amp;id=1425170"><em>Homo
heidelbergensis</em></a> and <a href="https://www.ncbi.nlm.nih.gov/Taxonomy/Browser/wwwtax.cgi?mode=Tree&amp;id=9606"><em>Homo
sapiens</em></a> is the genus <a href="https://www.ncbi.nlm.nih.gov/Taxonomy/Browser/wwwtax.cgi?mode=Undef&amp;id=9605"><em>Homo</em></a>.
Note that LCA is a property of a graph, in computer science, but i opted
to organize this as a data.table, as recovering LCA was faster (for me
at least, using the R tools I tried) in this format than in graph
format.</p>
<p>To build the cid.lca relationships, we need to define which ‘sources’
of taxonomy information we wish to use. for this example, we will just
use LOTUS, but there are many options. I will discuss the importance of
this below, but for now, here is the code we can use to infer LCA for
each CID.</p>
<div id="note-that-running-this-line-of-code-may-will-take-about-an-hour-of-computer-time-with-8-threads-64-gb-ram" class="section level6">
<h6>note that running this line of code may will take about an hour of
computer time, with 8 threads, 64 GB RAM</h6>
<div class="sourceCode" id="cb3"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb3-1"><a href="#cb3-1" tabindex="-1"></a><span class="fu">build.cid.lca</span>(<span class="at">pc.directory =</span> <span class="st">&quot;C:/Temp/20250703&quot;</span>, <span class="at">tax.sources =</span>  <span class="st">&quot;LOTUS - the natural products occurrence database&quot;</span>)</span></code></pre></div>
<p>This function builds the CID - LCA relationship for all compounds
with taxonomy data. Note that you can optionally use pathway information
in your taxonomy source data as well, as many of them do have taxonmy
assigned. In the case of pathway data, there are conserved and species
specific pathways. Conserved pathways are considered to have an LCA = 1,
or ‘root’ taxonomy, and therefore can be used to ensure that the TCA
cycle metabolites, for example, are going to be included in any
metabolomic database generated (they would be anyway - this is a trivial
example), even if there is no specific evidence that it is present in
your species of interest. By default ‘use.conserved.pathways’ is set to
‘FALSE’, as ‘conserved’ can mean different things to different people.
If you set ‘use.conserved.pathways = TRUE’, LCA = 1 for every metabolite
mentioned in a pathway which is listed as ‘conserved’.</p>
<p>This function is not building your taxonomy specific database, it is
assigning a lowest common ancestory for every metabolite with supporting
taxonomy data. Ultimately, we will build the taxonomy specific databases
from the output of this function, for any taxa we like. Ultimately, you
will still need to demonstrate, using analytical mass spectrometry, that
capsaicin is actually present in the species (or not) - we are just
defining chemical search space. As such, I tend to approach this in an
‘inclusive’ manner - if a metabolite is plausibly present, i want it in
the search space.</p>
<p>Below is a subset of the taxid.hierarchy file. This gives you an idea
of how the LCA’s are assigned.</p>
<div class="sourceCode" id="cb4"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb4-1"><a href="#cb4-1" tabindex="-1"></a><span class="fu">data</span>(sub.taxid.hierarchy, <span class="at">package =</span> <span class="st">&#39;pubchem.bio&#39;</span>)</span>
<span id="cb4-2"><a href="#cb4-2" tabindex="-1"></a>sub.taxid.hierarchy</span></code></pre></div>
<pre><code>##   species genus subtribe  tribe subfamily family superfamily infraorder
## 1    1173  1163       NA     NA        NA   1162          NA         NA
## 2    4072  4071       NA 424564    424551   4070          NA         NA
## 3    4081  4107       NA 424574    424551   4070          NA         NA
## 4    4232  4231       NA 102814    102804   4210          NA         NA
## 5    4932  4930       NA     NA        NA   4893          NA         NA
##   suborder order superorder infraclass subclass   class superclass subphylum
## 1       NA  1161         NA         NA       NA 3028117         NA        NA
## 2       NA  4069         NA         NA       NA    3398         NA    131221
## 3       NA  4069         NA         NA       NA    3398         NA    131221
## 4       NA  4209         NA         NA       NA    3398         NA    131221
## 5       NA  4892         NA         NA       NA    4891         NA    147537
##   phylum subkingdom kingdom superkingdom root
## 1   1117         NA 1783272           NA    1
## 2  35493         NA   33090           NA    1
## 3  35493         NA   33090           NA    1
## 4  35493         NA   33090           NA    1
## 5   4890     451864    4751           NA    1</code></pre>
<p>We have five taxa represented here: species 1173 = <em>Mycobacterium
tuberculosis</em> (Tuberculosis bacterium), 4072 = <em>Capsicum
annuum</em> (pepper), 4081 = <em>Solanum lycopersicum</em> (tomato),
4232 = <em>Helianthus annuus</em> (sunflower), 4932 = <em>Saccharomyces
cerevisiae</em> (brewers yeast). Tomato and Pepper are the two most
closely related species, and you can see this in the table above by
looking for the left-most common value in those two rows - while
‘species’, ‘genus’, and ‘tribe’ are different, they share the same
‘family’ - taxid <a href="https://www.ncbi.nlm.nih.gov/Taxonomy/Browser/wwwtax.cgi?mode=Info&amp;id=4070">4070</a>.
Sunflower, also a vascular plant, but more distantly related, does not
share a taxid with tomato/pepper until ‘class’. Mycobacterium and yeast
do not match until ‘root.’ Taxonomy is therefore hierarchical, and we
can map our CID - LCA relationships onto it.</p>
<p>There are dedicated taxonomy-cid relationships defined by many
groups, or in PubChem vernacular, ‘sources’. For example, if we look at
the taxonomic relationships for capsaicin (see URL above), we see that
it is noted to be found in various <em>Capsicum</em> species, which is
expected, but it is also noted to be found in humans through its
presence in HMDB. This record for humans is certainly true, in one sense
- we humans often eat spicy peppers and therefore Capsaicin will be
found in humans. But it is not <em>produced</em> by humans.</p>
<p>It is up to you whether you want to capture these sorts of
relationships in your taxonomy data, and you have to impliment that
decision by choosing data sources appropriately. I have set the
‘tax.sources’ option default to be ‘interactive’ so that every time you
can see the sources and decide. However, you can also just set the
sources as a character vector. i.e. c(“LOTUS - the natural products
occurrence database”) will ONLY use taxonomy-CID relationships as
defined by LOTUS. Also note that some sources appear to have what i
would consider to be false positive assignments. Capsicum is reported to
be found in <em>Mango</em> by the NPASS database. This appears to be a
product of automated text mining mis-assignment. Given that LOTUS seems
to contain fewer of these (i have not been systematic or comprehensive
in evaluating this), and is also massive, i generally <em>only</em> use
LOTUS, but this is ulitmately up to you to decide.</p>
<p>As an example of why this all matters: If you set ‘tax.sources =
c(“LOTUS - the natural products occurrence database”)’, the LCA for
Capsaicin is <em>Capsicum</em>, which make great sense, since this is
the genus which contains all of the peppers we know and love. However,
if you use ‘NPASS’ in your ‘tax.sources’ option, the lowest common
ancestor for Capsacin changes from <em>Capsicum</em> to Petapetalae, a
clade containing all Asterids, Caryophyllales, Rosids, etc. This would
mean that if you built a species specific database for <em>Glycine
max</em>, the common soybean, it <em>would</em> contain Capsaicin, since
it falls within the taxonomy structure headed by the LCA Pentapetalae:
Pentapetalae/rosids/fabids/Fabales/Fabaceae/Papilionoideae/Phaseoleae/Glycine/max.</p>
</div>
</div>
<div id="creating-pubchem-bio-metabolome-database." class="section level3">
<h3>Creating PubChem BIO metabolome database.</h3>
<p>We can now build a database of biological compounds as a subset of
PubChem data. We base this on (1) BIO databases as a source (2) pathway
data and (3) taxonomy data. If you would like to see the full list of
all data sources, please go to PubChem: <a href="https://pubchem.ncbi.nlm.nih.gov/sources/" class="uri">https://pubchem.ncbi.nlm.nih.gov/sources/</a>. Currently,
the default uses bio.sources = c(“Metabolomics Workbench”,“Human
Metabolome Database (HMDB)”, “ChEBI”, “LIPID MAPS”, “MassBank of North
America (MoNA)”). All pathway metabolites are added, as are all
metabolites from the taxonomy dataset. You can specify to only include
certain sources for each of pathways and taxonomy as well. Generally,
when compiling metabolomics databases, we want to remove salts as they
are not detected in salt form by LC- or GC-MS, so that is done by
default. We also can optionally calculate some basic physicochemical
properties using the ‘rcdk’ package. While some of these physicochemical
properties are also available via PubChem, i have found it much faster
to calculate them locally than to retrieve them from PubChem, and
building the rcdk property calculation into the function at this stage
makes the package output more versatile, as there are <em>many</em> rcdk
properties which are not available via PubChem. Let me know if you feel
it important to change this behavior and actually use PubChem predicted
properties instead.</p>
<div id="note-that-running-this-line-of-code-may-will-take-about-1.5-hours-of-computer-time-with-8-threads-64-gb-ram" class="section level6">
<h6>note that running this line of code may will take about 1.5 hours of
computer time, with 8 threads, 64 GB RAM</h6>
<div class="sourceCode" id="cb6"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb6-1"><a href="#cb6-1" tabindex="-1"></a>pc.bio <span class="ot">&lt;-</span> <span class="fu">build.pubchem.bio</span>(<span class="at">pc.directory =</span> <span class="st">&quot;C:/Temp/20250703&quot;</span>)</span></code></pre></div>
<p>This ‘pc.bio’ object contains &gt; 1M metabolites (using default
settings), and has columns with names, structures, and properties. It
can be used directly as a source dataset for export to spreadsheet
format, as it is a data.frame object. It includes metadata including the
number of references to pubmed (‘pmid.ct’), the number of sources that
include this compound in their PubChem submission (‘source.ct’), number
of pathways that reference the compound (‘pathway.ct’) and the number of
taxonomy relationships defined (‘taxonomy.ct’). These are only reported
- it is up to you what to do with this information. Some informatics
tools rely heavily on such information to increase the true positive
hits, while others vehemently eschew such approaches as it introduces
bias into the analysis. Additionally, we have now four rcdk-calculated
properties, ‘XLogP’, ‘nAcid’, ‘nBase’, and ‘TopoPSA’. These properties
are useful entry points into chemical descriptions that are of value in
undersatnding analytical behavior. ‘XLogP’ as an estimator of neutral pH
hydrophobicity, ‘nAcid’ and ‘nBase’ to describe ionizability, and
‘TopoPSA’ as a tool for understanding polarity in a manner dependent on
molecular size. Such properties can be valuable in predicting retention
time, CCS, adduct affinity, etc.</p>
</div>
</div>
<div id="creating-taxon-specific-metabolite-databases" class="section level3">
<h3>Creating taxon-specific metabolite databases</h3>
<p>By organizing taxonomy-metabolite data within the taxonomic hierarchy
in CID - LCA pairs (above), we can use some logical/reasoned steps to
ensure that our taxonomy-specific metabolome database is as
comprehensive as possible. We know that <a href="https://pubchem.ncbi.nlm.nih.gov/compound/Capsaicin">Capsaicin</a>
is found in <em>Capsicum annuum</em> and <em>Capsicum frutescens</em>
(amongst others). If the only two know sources of Capsacin were the two
<em>Capsicum</em> species, we can set the lowest common ancestor for
Capasaicin to the <em>Capsicum</em> genus. Additionally, this allows us
to <em>reasonably</em> infer that capsaicin <em>could</em> be found in
other <em>Capsicum</em> species, such as <em>Capsicum cornutum</em> even
if there is no documented evidence of this in PubChem. Do note that this
inference is not proof of capsaicin is found in all <em>Capsicum</em>
species, but the data supports to notion that it is quite plausible that
capsaicin may be present in other <em>Capsicum</em> species as well.
Evidence-based inference. This can be inferred at every taxonomic level.
The more taxonomy associations are present in PubChem, the more data we
have to infer the taxonomic breadth of the metabolic process that can
produce that metabolite. Some metabolic pathways can be considered to be
‘conserved’ - found in all known biology. The TCA cycle, glycolysis,
fatty acid biosynthesis, etc. The data in PubChem enable us to assign an
LCA of ‘1’ - the root of all biology - to such metabolite CIDs. These
metabolites will also therefore be expected to be part of the
<em>Capsicum cornutum</em> metabolome, since it is nested under the
‘root’ taxonomic hierarchy.</p>
<p>This ultimately enables us to build a metabolome database for
<em>Capsicum cornutum</em> even if PubChem had no information on
<em>Capsicum cornutum</em> metabolites, since we have defined many
metabolites which associate with the <em>Capsicum</em> genus, as well as
with higher taxonomic levels.</p>
<div id="running-this-line-of-code-may-will-take-about-a-minute-of-computer-time-with-8-threads-64-gb-ram" class="section level6">
<h6>running this line of code may will take about a minute of computer
time, with 8 threads, 64 GB RAM</h6>
<div class="sourceCode" id="cb7"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb7-1"><a href="#cb7-1" tabindex="-1"></a>pc.bio.sub <span class="ot">&lt;-</span> <span class="fu">build.taxon.metabolome</span>(<span class="at">taxid =</span> <span class="dv">1710960</span>, <span class="at">pc.directory =</span> <span class="st">&quot;C:/Temp/20250703&quot;</span>, <span class="at">get.properties =</span> <span class="cn">FALSE</span>)</span></code></pre></div>
</div>
</div>
<div id="meta-metabolomes" class="section level3">
<h3>Meta-metabolomes</h3>
<p>This function can also take a list of taxids. Imagine you are
analyzing stool samples from a gut microbiome study under controlled
diet conditions. You can include human (9606), <em>Prevotella</em>
(838), <em>Bacterioides</em> (816), <em>Capsicum</em> (4071), <em>Bos
taurus</em> (9913), etc, by setting taxid = c(9606, 838, 816, 4071,
9913). This tool does not try to predict cross-species metabolism, but
you now have a metabolite database which you can provide to tools such
as <a href="https://biotransformer.ca/">BioTransformer</a>.</p>
<p>Let’s consider a meta-metabolome that is a bit more fun. Salsa. We
can simplify the ingredient list a bit. lets consider peppers, tomatoes,
and cilantro.</p>
<div class="sourceCode" id="cb8"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb8-1"><a href="#cb8-1" tabindex="-1"></a>pc.bio.sub <span class="ot">&lt;-</span> <span class="fu">build.taxon.metabolome</span>(<span class="at">taxid =</span> <span class="fu">c</span>(<span class="dv">4072</span>, <span class="dv">4107</span>, <span class="dv">4047</span>), <span class="at">pc.directory =</span> <span class="st">&quot;C:/Temp/20250703&quot;</span>, <span class="at">get.properties =</span> <span class="cn">FALSE</span>)</span></code></pre></div>
<p>In this example, I used <em>Capsicum annuum</em> (pepper, 4072), the
genus <em>Solanum</em> (tomato, 4107), and <em>Coriandrum sativum</em>
(cilantro/coriander, 4047) as the taxa used to construct this
meta-metabolome. Note that i used <em>Solanum</em> - a genus - rather
than <em>Solanum lycopersicon</em> - a species. This is fine. you can
always do this. You can use any taxonomic level in building your
metabolome (or meta-metabolome). While it may not make perfect sense to
do so, I wanted to demonstrate that you can.</p>
</div>
<div id="taxon-scored-metabolite-database-structure" class="section level3">
<h3>Taxon-scored metabolite database structure</h3>
<p>When we created the pc.bio object above, we also saved a version of
it to the pc.directory specified. We load this into memory, and then use
the mapped CID - LCA data, the taxonomy hierarchy, to (1) determine all
metabolites whose LCA maps to any taxonomic level in the taxa we ask and
(2) assign a taxonomy-LCA score to all metabolite CIDs which have any
taxonomy data. All CIDs which map to the taxa listed are assinged a
taxonomy.lca.similarity score of 1. All metabolites with taxonomy data
but are not mapped to the selected taxon are assigned a value greater
than or equal to zero and less than one. The closer the taxonomic level
of the CID’s LCA is to the lowest common ancestor of the mapping taxon,
the higher the similarity score. There are 21 levels to the taxonomic
hierarchy. If a CID maps to species A only, and we ask for species B’s
metabolome, and A and B are member of the same genus, then there is one
taxonomic level between the two levels, so the CID is assigned a
taxonomy.lca.similarity of (1 - 1/21) = 0.9523. A value near, but less
than, 1.</p>
<p>Let’s spend a bit of time looking at a small subset of the ouptut of
our salsa meta-metabolome.</p>
<div class="sourceCode" id="cb9"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb9-1"><a href="#cb9-1" tabindex="-1"></a><span class="fu">data</span>(pc.bio.subset, <span class="at">package =</span> <span class="st">&#39;pubchem.bio&#39;</span>)</span>
<span id="cb9-2"><a href="#cb9-2" tabindex="-1"></a>pc.bio.subset[,<span class="fu">c</span>(<span class="st">&quot;cid&quot;</span>, <span class="st">&quot;name&quot;</span>, <span class="st">&quot;taxonomy.lca.similarity.4072&quot;</span>, <span class="st">&quot;taxonomy.lca.similarity.4107&quot;</span>, <span class="st">&quot;taxonomy.lca.similarity.4047&quot;</span>, <span class="st">&quot;taxonomy.lca.similarity.aggregate&quot;</span>)]</span></code></pre></div>
<pre><code>##                 cid                 name taxonomy.lca.similarity.4072
## 311             311          Citric Acid                            1
## 174174       174174             Atropine                            1
## 1548943     1548943            Capsaicin                            1
## 21585658   21585658           Daptomycin                            0
## 139590519 139590519 Saccharomonopyrone C                            0
##           taxonomy.lca.similarity.4107 taxonomy.lca.similarity.4047
## 311                             1.0000                       1.0000
## 174174                          1.0000                       0.3333
## 1548943                         0.7619                       0.3333
## 21585658                        0.0000                       0.0000
## 139590519                       0.0000                       0.0000
##           taxonomy.lca.similarity.aggregate
## 311                                       1
## 174174                                    1
## 1548943                                   1
## 21585658                                  0
## 139590519                                 0</code></pre>
<p>We defined our target taxa as pepper (species), the tomato genus, and
the cilantro species. Each taxon is assigned its own similarity score
(with the taxid appended to ‘taxonomy.lca.similarity’). For citric acid,
a highly conserved metabolite, all taxa are assigned a similarity score
of 1. Also enabled is an aggregation function - the default being ‘max’.
Citric acid’s aggregate taxonomy.lca.similarlity score is also 1.
Capsaicin is specific to peppers. We can see that the assigned
similarity score for peppers is 1 for taxid 4072 (peppers), 0.76 for
4107 (tomato genus), and 0.33 for 4047 (cilantro). Since cilantro is
more distantly related to pepper than tomato, which occupies the same
solanceous clade, it is much less likely that capsaicin would be
produced by cilantro - this is reflected in the scoring. Daptomycin is
produced by <em>Streptomyces filamentosus</em> and Saccharamonopyrone C
by brewers yeast, <em>Saccharomyces cerevisiae</em> - these lowest
common ancestor between pepper (or tomato, or cilantro) to these species
is ‘root’, since they are fungal species. As such, it is extremly
unlikely that these compounds, which are only known to occur from one
species (in PubChem, at least), are going to be found in our salsa - the
scores are assigned as 0. Atropine is found in a few different
solanaceous species (i.e. <em>Duboisia spp</em>, <em>Atropa spp</em>,
<em>Datura spp</em>, etc). It is not reported in tomato or pepper, but
since its LCA is assigned as 4070 - <a href="https://www.ncbi.nlm.nih.gov/Taxonomy/Browser/wwwtax.cgi?mode=Info&amp;id=4070&amp;lvl=3&amp;lin=f&amp;keep=1&amp;srchmode=1&amp;unlock">Solanaceae</a>,
atropine is considered to plausibly be found in all <em>Solanaceae</em>
species, and assigned scores of 1 for each of 4072 and 4107. This
example demonstrates both the benefit of the LCA approach - Citric acid
and all other highly conserved metabolites are mapped to every species -
but also demonstrates a cautionary tale - atropine is mapped to all
Solanaceous species, even though there is no evidence that it is likely
to be found in peppers and tomatoes, or our salsa. You should be trying
to think about this in evolutionary terms, and there are two
explanations: (1) Atropine synthesis capacity arose randomly a few
genera, all of which happen to be in the <em>Solanacea</em> or (2) The
common ancestor to all <em>Solanacea</em> had genetic traits which
predisposes all of the progeny to develop the ability to synthesize
atropine. We can’t tell the difference, but we are using taxonomy to
infer phylogeny, and in this case, the impact of phylogeny on inferred
metabolic traits.</p>
<p>If you wish to create a metabolome <em>ONLY</em> with metabolites
which are plausible, you can subset the data to keep only those
metabolite CIDs with a similarilty score == 1. Something like:</p>
<p><code>my.metabolome &lt;- pc.bio.subset[which(pc.bio.subset$taxonomy.lca.similarity.4072 == 1),]</code></p>
<p>Alternatively, you can use the assigned scores to rank/rescore search
results from NIST MSsearch, Sirius, or any other search tool, so long as
you can map the data using CID, Smiles, InChIKey, etc. Note also that we
used ‘max’ as our aggregation function. You could alternatively use
‘mean’ or ‘min’ or any other function - even custom functions - to
aggregate this data. pubchem.bio is designed to do a good deal of the
basic data wrangling, using all available sources, but users of
pubchem.bio are going to be doing a good deal of the thinking as to how
best to use this resource.</p>
<p>One final note on taxonomy scoring: all CIDs from the pc.bio object
are included by default in the taxonomy scoring output. However, not all
CIDs have associated taxonomy information. Any CID which has no taxonomy
information (using your chosen CID - LCA parameters) will have a score
of NA. It will be up to you to determine how you want to handle
these.</p>
</div>
<div id="exporting-your-custom-database" class="section level3">
<h3>Exporting your custom database</h3>
<p>We provide three export options. In each case, we take as input the
pc.bio object or the related taxonomy-scored pc.bio object. At this
point, taxonomy scoring isn’t feasible in these programs - we are
exporting the minimal data asked for by the programs, and taxonomy score
is not one of them. We use the CID as the database ID in each case - no
point in creating another database ID when we have a perfectly good one
from PubChem. There is also a convenience wrapper to export the full
pubchem.bio table to .tsv format - export.pubchem.bio().</p>
<div class="sourceCode" id="cb11"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb11-1"><a href="#cb11-1" tabindex="-1"></a><span class="fu">export.msfinder</span>(<span class="at">pc.bio.object =</span> pc.bio.sub, <span class="at">export.file.name =</span> <span class="st">&quot;C:/Temp/20250703/pc.bio.msfinder.tsv&quot;</span>)</span>
<span id="cb11-2"><a href="#cb11-2" tabindex="-1"></a><span class="fu">export.sirius</span>(<span class="at">pc.bio.object =</span> pc.bio.sub, <span class="at">export.file.name =</span> <span class="st">&quot;C:/Temp/20250703/pc.bio.sirius.tsv&quot;</span>)</span>
<span id="cb11-3"><a href="#cb11-3" tabindex="-1"></a><span class="fu">export.pubchem.bio</span>(<span class="at">pc.bio.object =</span> pc.bio.sub, <span class="at">export.file.name =</span> <span class="st">&quot;C:/Temp/20250703/pc.bio.full.pubchem.bio.tsv&quot;</span>)</span></code></pre></div>
</div>
<div id="the-future" class="section level3">
<h3>The future</h3>
<p>if you look closely at the data that has been downloaded, you will
find a good deal of additional data - all of it mapped by CID. This
includes MESH names/functions, pubmed reference identifiers, synonyms,
pathway membership, etc. These data can ultimately be used for mapping
your pubchem.bio data to biology, enabling, for example, enrichment
analysis. I do not have all these data mapped in the output data.table
yet, but they are available for you to use in the interim.</p>
</div>
<div id="workflow-in-brief" class="section level3">
<h3>Workflow in brief</h3>
<p>Just make sure to set the local.pubchem.directory to something that
makes sense for your system.</p>
<div class="sourceCode" id="cb12"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb12-1"><a href="#cb12-1" tabindex="-1"></a><span class="fu">install.packages</span>(pubchem.bio, <span class="at">dependencies =</span> <span class="cn">TRUE</span>)</span>
<span id="cb12-2"><a href="#cb12-2" tabindex="-1"></a><span class="fu">library</span>(pubchem.bio)</span>
<span id="cb12-3"><a href="#cb12-3" tabindex="-1"></a>local.pubchem.directory <span class="ot">&lt;-</span> <span class="st">&quot;C:/Temp/20250801&quot;</span>  <span class="do">## i suggest this naming scheme, but feel free to chose your own.</span></span>
<span id="cb12-4"><a href="#cb12-4" tabindex="-1"></a>pubchem.bio<span class="sc">::</span><span class="fu">get.pubchem.ftp</span>(<span class="at">pc.directory =</span> local.pubchem.directory)</span>
<span id="cb12-5"><a href="#cb12-5" tabindex="-1"></a>cid.lca <span class="ot">&lt;-</span> pubchem.bio<span class="sc">::</span><span class="fu">build.cid.lca</span>(<span class="at">pc.directory =</span> local.pubchem.directory, <span class="at">tax.sources =</span>  <span class="st">&quot;LOTUS - the natural products occurrence database&quot;</span>)</span>
<span id="cb12-6"><a href="#cb12-6" tabindex="-1"></a>pc.bio <span class="ot">&lt;-</span> pubchem.bio<span class="sc">::</span><span class="fu">build.pubchem.bio</span>(<span class="at">pc.directory =</span> local.pubchem.directory)</span>
<span id="cb12-7"><a href="#cb12-7" tabindex="-1"></a>pc.bio.salsa <span class="ot">&lt;-</span> pubchem.bio<span class="sc">::</span><span class="fu">build.taxon.metabolome</span>(<span class="at">taxid =</span> <span class="fu">c</span>(<span class="dv">4081</span>, <span class="dv">4072</span>, <span class="dv">4047</span>, <span class="dv">4679</span>, <span class="dv">4682</span>), <span class="at">pc.directory =</span> local.pubchem.directory, <span class="at">get.properties =</span> <span class="cn">FALSE</span>)</span>
<span id="cb12-8"><a href="#cb12-8" tabindex="-1"></a><span class="fu">export.msfinder</span>(<span class="at">pc.bio.object =</span> pc.bio.salsa, <span class="at">export.file.name =</span> <span class="fu">paste0</span>(local.pubchem.directory, <span class="st">&quot;/pc.bio.msfinder.tsv&quot;</span>))</span>
<span id="cb12-9"><a href="#cb12-9" tabindex="-1"></a><span class="fu">export.sirius</span>(<span class="at">pc.bio.object =</span> pc.bio.salsa, <span class="at">export.file.name =</span> <span class="fu">paste0</span>(local.pubchem.directory, <span class="st">&quot;/pc.bio.sirius.tsv&quot;</span>))</span>
<span id="cb12-10"><a href="#cb12-10" tabindex="-1"></a><span class="fu">export.pubchem.bio</span>(<span class="at">pc.bio.object =</span> pc.bio.salsa, <span class="at">export.file.name =</span> <span class="fu">paste0</span>(local.pubchem.directory, <span class="st">&quot;/pc.bio.full.pubchem.bio.tsv&quot;</span>))</span></code></pre></div>
</div>
</div>



<!-- code folding -->


<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
